### PR TITLE
Custom certificates

### PR DESCRIPTION
**Pull request by etsauer**
#### What is this PR About?
Adds a guide to configuring custom certificates in OpenShift.

#### How should we test or review this PR?
Read through new guide, and changes to install guide.

#### Is there a relevant Trello card or Github issue open for this?
https://trello.com/c/r1Vj358a/5-1-custom-certificate-management-document

#### Who would you like to review this?
cc: @redhat-cop/cant-contain-this @timothyh

### CODE REVIEW

File: playbooks/installation/custom_certificates.adoc [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+:service_account: Service Account Auth via Client Certificate
+:raw_client_cert: Raw Client Certificate Auth
+:api_token: API Token over HTTPS connection
+:https: HTTPS Connection
+.OpenShift Certificate Usage
+[options="header"]
+|====================
+| Source (Client) | Destination (Server) | TLS Type | Notes
+| Nodes | Master API | {service_account} | Node Service account have the format `system:node:<hostname>`
```
Comment (juanluisvaladas) [inline playbooks/installation/custom_certificates.adoc]: Technically speaking that's not a service account.

Comment (juanluisvaladas) [inline playbooks/installation/custom_certificates.adoc]: https://kubernetes.io/docs/admin/authentication/#x509-client-certs

File: playbooks/installation/custom_certificates.adoc [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+| Master | Node (kubelet) | {raw_client_cert} |
+| `system:admin` User | Master API | {service_account} | This is the default cluster admin user. Only accessible as `root` on a master host, as this is the only user that has read access to the client certificate for the account.
```
Comment (juanluisvaladas) [inline playbooks/installation/custom_certificates.adoc]: Also not a service account, it's also an 509 client cert

File: playbooks/installation/custom_certificates.adoc [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+| Kibana | ElasticSearch | {raw_client_cert} |
+| User | Hawkular (Metrics) | {api_token} |
+| User | Registry | {api_token} |
+| User | Registry Console | {api_token} |
+| Registry Console | Registry | {service_account} |
+|====================
+
+To make the provisioning process for all of these certificates more palatable, the Ansible Config Playbook that is used to automate the deployment and configuration of OpenShift automatically generates and signs all of these certificates. By default, this is done from an internally signed CA (the `openshift-signer`). This is convenient, as it makes the creation, signing, and configuring of these certificates essentially invisible to the administrator, and the install is as seamless as possible. However, when building long running, higher environment OpenShift clusters, it is often desired to have certificates signed by a trusted internal or public Certificate Authority presented to users. The Ansible Config Playbook supports two options for handling custom certificates, covered in the following sections.
+
+=== Using a Trusted Intermediate CA
```
Comment (juanluisvaladas) [inline playbooks/installation/custom_certificates.adoc]: I'm not sure but I think this isn't supported. If it's not add a warning

File: playbooks/installation/custom_certificates.adoc [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+----
+ansible-playbook -i c1-ocp.myorg.com/hosts /usr/share/ansible/openshift-ansible/playbooks/byo/config.yml
+----
+
+=== Option 2: Configuring OpenShift to use Component-specific Custom Certificates
+
+Alternatively, you can allow OpenShift to use its own CA to generate the internal certificates (self-signed) with which all of the internal components of OpenShift will authenticate. Custom certificates may be created and configured to be used individually by User-facing components of openshift. Discussion of each component is below.
+
+INFO: One way to simplify this is to have one certificate with multiple SAN names (console.example.com, metrics.example.com, logging.example.com), and to use the same certificate for all components.
```
Comment (juanluisvaladas) [inline playbooks/installation/custom_certificates.adoc]: I don't like this, I prefer having a certificate per site. I would either use a wildcard and a reencrypt route for stuff like the registry and logging than having one for everything.

It's just an opinion and not a request for a change.

File: playbooks/installation/index.adoc [hunk 5]
```diff
@@ -127,41 +127,32 @@ More information on Pods & Services can be found in the link:https://docs.opensh
 
 All of the hosts in the cluster need to be resolveable via DNS.  Additionally if using a control node to serve as the ansible installer it too should be able to resolve all hosts in your cluster.
 
-In an HA cluster there will also be a DNS name for the Load Balanced URL that points to the 3 master servers for access to the API, CLI and Console services.  This value should also resolve, and will be placed in the ansible Hosts file for the variables.
+In an HA cluster there should also be two DNS names for the Load Balanced IP address that points to the 3 master servers for access to the API, CLI and Console services. One of these names is the *public* name that users will use to log into the cluster. The other is an _internal_ name that will be used by service accounts within the cluster to talk back to the master. These values should also resolve, and will be placed in the ansible Hosts file for the variables.
```
Comment (juanluisvaladas) [inline playbooks/installation/index.adoc]: s/service accounts/internal components/

File: playbooks/installation/custom_certificates.adoc (hunk 4 above)
Comment (etsauer) [inline playbooks/installation/custom_certificates.adoc]: I'm fine not including it.

File: playbooks/installation/custom_certificates.adoc (hunk 3 above)
Comment (sdodson) [inline playbooks/installation/custom_certificates.adoc]: It's supported and has been tested by QE.

File: playbooks/installation/index.adoc [hunk 6]
```diff
@@ -127,41 +127,32 @@ More information on Pods & Services can be found in the link:https://docs.opensh
...
 
-This will be the url that external users and/or tools will use to login to the OpenShift cluster
-
-Ansible Host file variable:
-
-*Master Public Hostname Variable :* `openshift_master_cluster_public_hostname=console.c1-ocp.myorg.com`
+*Ansible Inventory Variable:* `openshift_master_cluster_public_hostname=console.c1-ocp.myorg.com`
 
-Replace the value in your host file
+This will be the url that external users and/or tools will use to login to the OpenShift cluster
```
Comment (newgoliath) [inline playbooks/installation/index.adoc]: s/url/hostname/

File: playbooks/installation/index.adoc [hunk 7]
```diff
@@ -127,41 +127,32 @@ More information on Pods & Services can be found in the link:https://docs.opensh
...
 
+==== Internal Master Hostname
 
-Master URL
+*Ansible Inventory Variable:* `openshift_master_cluster_hostname=console-int.c1-ocp.myorg.com`
 
 This will be the url that external users and/or tools will use to login to the OpenShift cluster
```
Comment (newgoliath) [inline playbooks/installation/index.adoc]: s/url/hostname/

File: playbooks/installation/custom_certificates.adoc [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+OpenShift creates a lot of certificates by default. Understanding where and how those certificates are used will help to understand how it can be configured to meet your needs.
+
+In this guide we will discuss how and where TLS Certificates are used in OpenShift, how they are created, and how to go about designing a proper certificate management strategy for OpenShift. We will then show you how to configure OpenShift via Ansible to reflect this strategy.
+
+== Design of OpenShift Certificate Management
+
+OpenShift utilizes TLS Certificates in three primary ways.
+
+1. To encrypt traffic
+1. To Authenticate Certain Infrastructure Components
```
Comment (newgoliath) [inline playbooks/installation/custom_certificates.adoc]: normalize for sentence case.

File: playbooks/installation/custom_certificates.adoc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+1. To authenticate certain users
+  * OpenShift uses client certificates to authenticate system critical Service Accounts
```
Comment (newgoliath) [inline playbooks/installation/custom_certificates.adoc]: s/service accounts/system users and service accounts/

https://docs.openshift.com/container-platform/3.4/architecture/core_concepts/projects_and_users.html

File: playbooks/installation/custom_certificates.adoc (hunk 4 above)
Comment (newgoliath) [inline playbooks/installation/custom_certificates.adoc]: Just for reference: 

https://docs.openshift.com/container-platform/3.4/architecture/core_concepts/routes.html#secured-routes

**Comment by etsauer** [comment]
@newgoliath @juanluisvaladas updated based on your feedback. Please re-review.

File: playbooks/installation/index.adoc [hunk 10]
```diff
@@ -127,41 +127,32 @@ More information on Pods & Services can be found in the link:https://docs.opensh
...
-Ansible Host file variable:
+This will be the hostname that external users and/or tools will use to login to the OpenShift cluster
 
-*Master Public Hostname Variable :* `openshift_master_cluster_public_hostname=console.c1-ocp.myorg.com`
+==== Internal Master Hostname
 
-Replace the value in your host file
+*Ansible Inventory Variable:* `openshift_master_cluster_hostname=console-int.c1-ocp.myorg.com`
 
+This will be the hostname that external users and/or tools will use to login to the OpenShift cluster
```
Comment (oybed) [inline playbooks/installation/index.adoc]: Did you mean to say *internal* here?

File: playbooks/installation/index.adoc [hunk 11]
```diff
@@ -127,41 +127,32 @@ More information on Pods & Services can be found in the link:https://docs.opensh
...
-
-----
-# default subdomain to use for exposed routes
-openshift_master_default_subdomain=cloudapps.c1-ocp.myorg.com
-----
+In addition to the hostnames for the master console and API, A wildcard DNS entry needs to exist under a unique subdomain (i.e. `*.apps.c1-ocp.myorg.com`) that resolves to either the IP addresses (an A record) or the hostnames (a CNAME record) of the three Infrastructure Nodes.   This entry allows new routes to be automatically routable to the cluster under the subdomain such as mynewapp.apps.c1-ocp.myorg.com.   Alternatively, every exposed route would require the entry to be created in order to route it to the OpenShift cluster.
```
Comment (oybed) [inline playbooks/installation/index.adoc]: Maybe a note here pointing to the event-controller as an option (with the proper plugin enabled/added)?

File: playbooks/installation/index.adoc (hunk 10 above)
Comment (etsauer) [inline playbooks/installation/index.adoc]: @oybed I meant external to the cluster

**Comment by etsauer** [comment]
@oybed addressed your comments. If good, please go ahead and merge.

**Pull request merged**